Add torch._library.utils.normalize_args_kwargs

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148104
* #148092
* #148091
* #148063
* #148046
* __->__ #148062

Normalizes (args, kwargs) to the PyTorch dispatcher calling convention.
I need some sort of normalization in the next PR.

Test Plan:
- new tests

zou3519 committed:
Add torch._library.utils.normalize_args_kwargs

Normalizes (args, kwargs) to the PyTorch dispatcher calling convention.
I need some sort of normalization in the next PR.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch._library.utils.normalize_args_kwargs"


Normalizes (args, kwargs) to the PyTorch dispatcher calling convention.
I need some sort of normalization in the next PR.

Test Plan:
- new tests

[ghstack-poisoned]

Skylion007 commented on torch/_library/utils.py:
```diff
@@ -215,6 +215,67 @@ def zip_schema(
     return
 
 
+def normalize_args_kwargs(
```
Wait, I've definitely seen this function somewhere else in the recent inductor typing PRs, or atleast a very similar function. The args and kwargs should be annotated with ParamSpec so they they properly pass the types of the args and don't disable argument checking.

The other piece of code I reviewed recently was a decorator too.

zou3519 commented:
This function is different from torch.fx.operator_schema.normalize_function in that this function (1) only supports OpOverload and (2) this function removes default values where possible, which is closer to the PyTorch Dispatcher argument passing convention.

Skylion007 commented on torch/_library/utils.py:
```diff
@@ -215,6 +215,67 @@ def zip_schema(
 
 
+def normalize_args_kwargs(
+    schema: _C.FunctionSchema, args: tuple[Any, ...], kwargs: dict[str, Any]
```
You should annotate schema with ParamSpec so you can properly type the args and kwargs if you can. At the very least, the tuple here can be annotated with a TypeVarTuple to preserve the types within the tuple.

zou3519 committed:
Update on "Add torch._library.utils.normalize_args_kwargs"


Normalizes (args, kwargs) to the PyTorch dispatcher calling convention.
I need some sort of normalization in the next PR.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch._library.utils.normalize_args_kwargs"


Normalizes (args, kwargs) to the PyTorch dispatcher calling convention.
I need some sort of normalization in the next PR.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch._library.utils.normalize_args_kwargs"


Normalizes (args, kwargs) to the PyTorch dispatcher calling convention.
I need some sort of normalization in the next PR.

Test Plan:
- new tests

[ghstack-poisoned]
